src/components/container/SeverityCalendar/SeverityCalendar.tsx

```diff
@@ -0,0 +1,149 @@
+
+export interface SeverityCalendarProps {
+    surveyName: string,
+    dateRecordedResultIdentifier: string,
+    severityResultIdentifier: string,
+    intervalStart?: Date,
+    previewState?: SeverityCalendarPreviewState,
+}
+
+interface SeverityDay {
+    dateRecorded: string,
+    severity: string
+}
+
+export default function (props: SeverityCalendarProps) {
+    const surveyAnswerQuery: SurveyAnswersQuery = {
+        surveyName: props.surveyName,
+        resultIdentifier: [props.dateRecordedResultIdentifier, props.severityResultIdentifier]
+    }
+
+    const [data, setData] = useState<Map<string, SeverityDay> | undefined>();
+    const dateRangeContext = useContext(DateRangeContext);
+    let intervalStart = dateRangeContext?.intervalStart ?? props.intervalStart ?? startOfMonth(new Date());
+
+    async function initialize() {
+        if (props.previewState !== "Live") {
+            transformToCalendarData(props.previewState === "WithData" ? previewSeverityData : []);
+        } else {
+            surveyResults(surveyAnswerQuery).then((results: SurveyAnswer[]) => {
+                transformToCalendarData(results);
+            });
+        }
+    }
+
+    const stateConfiguration: CalendarDayStateConfiguration = {
+        'mild': {
+            style: {
+                background: '#F7DA19'
+            },
+            streak: true,
+            streakColor: '#FDEF97'
+        },
+        'moderate': {
+            style: {
+                background: 'var(--mdhui-color-warning)'
+            },
+            streak: true,
+            streakColor: '#FCCCC7'
+        },
+        'severe': {
+            style: {
+                background: 'var(--mdhui-color-danger)'
+            },
+            streak: true,
+            streakColor: '#FCCCC7'
+        },
+    };
+
+    const computeStateForDay = (date: Date): string => {
+        let key: string = startOfDay(date).toDateString();
+        return data?.get(key)?.severity ?? "";
+    };
+
+    function transformToCalendarData(results: SurveyAnswer[]) {
+        var severityEntries: Map<Guid, SeverityDay> = createResultsMap(results);
+
+        var calendarData: Map<string, SeverityDay> = new Map<string, SeverityDay>();
+        severityEntries.forEach(severityEntry => {
+            let key: string = startOfDay(parseISO(severityEntry.dateRecorded)).toDateString();
+            let exists = calendarData.get(key);
+            if (exists) {
+                if (parseISO(severityEntry.dateRecorded) > parseISO(exists.dateRecorded)) {
+                    exists.severity = severityEntry.severity;
+                }
+            }
+            else {
+                calendarData.set(key, severityEntry);
+            }
+        });
+
+        setData(calendarData);
+    }
+
+    function createResultsMap(results: SurveyAnswer[]) {
+        var groupedByResult = new Map<Guid, SeverityDay>();
+
+        results.forEach((r) => {
```

Comment:
I might do this a bit differently:
- First do a pass to group the survey answers by surveyResultID.  So you have an {[key:Guid]: SurveyAnswer[]}
- Then loop through that collection and:
  - If there's no severity answer all, skip this group
  - If there's no date answer at all, use the date from the severity answer (to allow for a situation where you aren't "dating" the logs" but are always logging for today)

In particular want to make sure this is relatively safe and if you misconfigure it, you don't end up with JS errors or anything...and it seems reasonable to fall back to using the date of the survey answer for severity

Reply:
This makes sense.